README.md
# senzing-package

## Overview

The [senzing-package.py](senzing-package.py) python script
copies `data/` and `g2/` directories to new locations.
The `senzing/senzing-package` docker image is a wrapper for use in docker formations (e.g. docker-compose, kubernetes).
Baked into `senzing/senzing-package` is the installed `senzingdata-v1` and `senzingapi` packages.

The dockerized version, `store/senzing/senzing-package`, is at
[hub.docker.com/_/senzing-package](https://hub.docker.com/_/senzing-package).
For more information, scroll down to [Accept docker image](#accept-docker-image).

To see all of the subcommands, run:

```console
$ ./senzing-package.py --help
usage: senzing-package.py [-h]
                          {install,package-version,sleep,version,docker-acceptance-test}
                          ...

Install Senzing packages. For more information, see
https://github.com/Senzing/senzing-package

positional arguments:
  {install,package-version,sleep,version,docker-acceptance-test}
                        Subcommands (SENZING_SUBCOMMAND):
    install             Copy source data and g2 directories to a target.
    package-version     Display Senzing Engine version inside docker image.
    sleep               Do nothing but sleep. For Docker testing.
    version             Print version of program.
    docker-acceptance-test
                        For Docker acceptance testing.

optional arguments:
  -h, --help            show this help message and exit
```

To see the options for a subcommand, run commands like:

```console
./senzing-package.py install --help
```

### Related artifacts

1. [DockerHub](https://hub.docker.com/_/senzing-package)
1. [Helm Chart](https://github.com/Senzing/charts/tree/master/charts/senzing-package)

### Contents

1. [Expectations](#expectations)
    1. [Space](#space)
    1. [Time](#time)
    1. [Background knowledge](#background-knowledge)
1. [Demonstrate using Docker](#demonstrate-using-docker)
    1. [Accept docker image](#accept-docker-image)
    1. [Configuration](#configuration)
    1. [Docker user](#docker-user)
    1. [Run docker container](#run-docker-container)
1. [Develop](#develop)
    1. [Prerequisite software](#prerequisite-software)
    1. [Clone repository](#clone-repository)
    1. [Build docker image for development](#build-docker-image-for-development)
1. [Examples](#examples)
1. [Errors](#errors)
1. [References](#references)

## Expectations

### Space

This repository and demonstration require 6 GB free disk space.

### Time

Budget 40 minutes to get the demonstration up-and-running, depending on CPU and network speeds.

### Background knowledge

This repository assumes a working knowledge of:

1. [Docker](https://github.com/Senzing/knowledge-base/blob/master/WHATIS/docker.md)

## Demonstrate using Docker

### Accept docker image

The use of the [store/senzing/senzing-package](https://hub.docker.com/_/senzing-package)
docker image requires acceptance of an End User License agreement (EULA).
To accept the license:

1. Visit [hub.docker.com/_/senzing-package](https://hub.docker.com/_/senzing-package).
1. Login to DockerHub.
1. On [hub.docker.com/_/senzing-package](https://hub.docker.com/_/senzing-package), click "Proceed to Checkout" button.
1. Check boxes for agreeing and acknowledging
1. Click "Get Content" button.
1. Verify image can be pulled from "Docker Store".  Example:

   ```console
   sudo docker pull store/senzing/senzing-package:0.0.1
   ```

### Configuration

Configuration values specified by environment variable or command line parameter.

- **[SENZING_ACCEPT_EULA](https://github.com/Senzing/knowledge-base/blob/master/lists/environment-variables.md#senzing_accept_eula)**
- **[SENZING_DATA_DIR](https://github.com/Senzing/knowledge-base/blob/master/lists/environment-variables.md#senzing_data_dir)**
- **[SENZING_DEBUG](https://github.com/Senzing/knowledge-base/blob/master/lists/environment-variables.md#senzing_debug)**
- **[SENZING_G2_DIR](https://github.com/Senzing/knowledge-base/blob/master/lists/environment-variables.md#senzing_g2_dir)**
- **[SENZING_LOG_LEVEL](https://github.com/Senzing/knowledge-base/blob/master/lists/environment-variables.md#senzing_log_level)**
- **[SENZING_NETWORK](https://github.com/Senzing/knowledge-base/blob/master/lists/environment-variables.md#senzing_network)**
- **[SENZING_RUNAS_USER](https://github.com/Senzing/knowledge-base/blob/master/lists/environment-variables.md#senzing_runas_user)**
- **[SENZING_SLEEP_TIME](https://github.com/Senzing/knowledge-base/blob/master/lists/environment-variables.md#senzing_sleep_time)**
- **[SENZING_SUBCOMMAND](https://github.com/Senzing/knowledge-base/blob/master/lists/environment-variables.md#senzing_subcommand)**

1. To determine which configuration parameters are use for each `<subcommand>`, run:

    ```console
    ./senzing-package.py <subcommand> --help
    ```

1. :pencil2: Set environment variables.  Example:

    ```console
    export SENZING_SUBCOMMAND=install
    export SENZING_DATA_DIR=~/my-senzing/data
    export SENZING_G2_DIR=~/my-senzing/g2
    ```

1. Create directories.
   Example:

    ```console
    mkdir -p ${SENZING_DATA_DIR}
    mkdir -p ${SENZING_G2_DIR}
    ```

### Docker user

:thinking: **Optional:**  The docker container runs as "USER root" by default.
Use if a different userid (UID) is required.

1. :pencil2: Manually identify user.
   Example:

    ```console
    export SENZING_RUNAS_USER="1001"
    export SENZING_RUNAS_GROUP="1001"
    ```

   Another option, use current user.
   Example:

    ```console
    export SENZING_RUNAS_USER=$(id -u)
    export SENZING_RUNAS_GROUP=$(id -g)
    ```

1. Construct parameter for `docker run`.
   Example:

    ```console
    export SENZING_RUNAS_USER_PARAMETER="--user ${SENZING_RUNAS_USER}:${SENZING_RUNAS_GROUP}"
    ```

### Run docker container

1. Run docker container.
   Example:

    ```console
    sudo docker run \
      ${SENZING_RUNAS_USER_PARAMETER} \
      --env SENZING_SUBCOMMAND="${SENZING_SUBCOMMAND}" \
      --rm \
      --volume ${SENZING_DATA_DIR}:/opt/senzing/data \
      --volume ${SENZING_G2_DIR}:/opt/senzing/g2 \
      senzing/senzing-package
    ```

## Develop

### Prerequisite software

The following software programs need to be installed:

1. [git](https://github.com/Senzing/knowledge-base/blob/master/HOWTO/install-git.md)
1. [make](https://github.com/Senzing/knowledge-base/blob/master/HOWTO/install-make.md)
1. [docker](https://github.com/Senzing/knowledge-base/blob/master/HOWTO/install-docker.md)

### Clone repository

For more information on environment variables,
see [Environment Variables](https://github.com/Senzing/knowledge-base/blob/master/lists/environment-variables.md).

1. Set these environment variable values:

    ```console
    export GIT_ACCOUNT=senzing
    export GIT_REPOSITORY=senzing-package
    export GIT_ACCOUNT_DIR=~/${GIT_ACCOUNT}.git
    export GIT_REPOSITORY_DIR="${GIT_ACCOUNT_DIR}/${GIT_REPOSITORY}"
    ```

1. Follow steps in [clone-repository](https://github.com/Senzing/knowledge-base/blob/master/HOWTO/clone-repository.md) to install the Git repository.

### Build docker image for development

To use any of the following options to build the Docker image which embeds Senzing code,
you must agree to the Senzing End User License Agreement (EULA).

:warning: This step is intentionally tricky and not simply copy/paste.
This ensures that you make a conscious effort to accept the EULA.
See
[SENZING_ACCEPT_EULA](https://github.com/Senzing/knowledge-base/blob/master/lists/environment-variables.md#senzing_accept_eula)
for the correct value.
Replace the double-quote character in the example with the correct value.
The use of the double-quote character is intentional to prevent simple copy/paste.
Example:

```console
export SENZING_ACCEPT_EULA="
```

1. **Option #1:** Using `docker` command and GitHub.

    ```console
    sudo docker build \
      --env SENZING_ACCEPT_EULA=${SENZING_ACCEPT_EULA} \
      --tag senzing/senzing-package \
      https://github.com/senzing/senzing-package.git
    ```

1. **Option #2:** Using `docker` command and local repository.

    ```console
    cd ${GIT_REPOSITORY_DIR}
    sudo docker build \
      --build-arg SENZING_ACCEPT_EULA=${SENZING_ACCEPT_EULA} \
      --tag senzing/senzing-package \
      .
    ```

1. **Option #3:** Using `make` command.

    ```console
    cd ${GIT_REPOSITORY_DIR}
    sudo \
      SENZING_ACCEPT_EULA=${SENZING_ACCEPT_EULA}
      make docker-build
    ```

    Note: `sudo make docker-build-development-cache` can be used to create cached docker layers.

## Examples

## Errors

1. See [docs/errors.md](docs/errors.md).

## References


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
[markdownlint](https://dlaa.me/markdownlint/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.12.1] - 2019-11-01

### Added to 1.12.1

- Bake in Senzing 1.12.1


rootfs/app/signal-trap.py
#! /usr/bin/env python3

# -----------------------------------------------------------------------------
# signal-trap.py  Test to see what signals are sent.
# -----------------------------------------------------------------------------

import signal
import sys
import time

signals = {
     "1": "SIGHUP",
     "2": "SIGINT",
     "3": "SIGQUIT",
     "4": "SIGILL",
     "5": "SIGTRAP",
     "6": "SIGABRT",
     "7": "SIGBUS",
     "8": "SIGFPE",
     "9": "SIGKILL",
     "10": "SIGUSR1",
     "11": "SIGSEGV",
     "12": "SIGUSR2",
     "13": "SIGPIPE",
     "14": "SIGALRM",
     "15": "SIGTERM",
     "16": "SIGSTKFLT",
     "17": "SIGCHLD",
     "18": "SIGCONT",
     "19": "SIGSTOP",
     "20": "SIGTSTP",
     "21": "SIGTTIN",
     "22": "SIGTTOU",
     "23": "SIGURG",
     "24": "SIGXCPU",
     "25": "SIGXFSZ",
     "26": "SIGVTALRM",
     "27": "SIGPROF",
     "28": "SIGWINCH",
     "29": "SIGIO",
     "30": "SIGPWR",
     "31": "SIGSYS",
     "34": "SIGRTMIN",
     "35": "SIGRTMIN+1",
     "36": "SIGRTMIN+2",
     "37": "SIGRTMIN+3",
     "38": "SIGRTMIN+4",
     "39": "SIGRTMIN+5",
     "40": "SIGRTMIN+6",
     "41": "SIGRTMIN+7",
     "42": "SIGRTMIN+8",
     "43": "SIGRTMIN+9",
     "44": "SIGRTMIN+10",
     "45": "SIGRTMIN+11",
     "46": "SIGRTMIN+12",
     "47": "SIGRTMIN+13",
     "48": "SIGRTMIN+14",
     "49": "SIGRTMIN+15",
     "50": "SIGRTMAX-14",
     "51": "SIGRTMAX-13",
     "52": "SIGRTMAX-12",
     "53": "SIGRTMAX-11",
     "54": "SIGRTMAX-10",
     "55": "SIGRTMAX-9",
     "56": "SIGRTMAX-8",
     "57": "SIGRTMAX-7",
     "58": "SIGRTMAX-6",
     "59": "SIGRTMAX-5",
     "60": "SIGRTMAX-4",
     "61": "SIGRTMAX-3",
     "62": "SIGRTMAX-2",
     "63": "SIGRTMAX-1",
     "64": "SIGRTMAX",
}


def handle_signal(signal, frame):
    print("\nSignal: {0} - {1}".format(signal, signals.get(str(signal))))
    sys.exit(0)

# -----------------------------------------------------------------------------
# Main
# -----------------------------------------------------------------------------


if __name__ == "__main__":

    for key, value in list(signals.items()):
        try:
            signal.signal(key, handle_signal)
        except:
            pass

    while True:
        print("Sleeping")
        time.sleep(3600)



senzing-package.py
#! /usr/bin/env python3

# -----------------------------------------------------------------------------
# senzing-package.py handles installation of the Senzing package.
# -----------------------------------------------------------------------------

import argparse
import distutils.dir_util
import grp
import json
import logging
import os
import pwd
import shutil
import signal
import sys
import tarfile
import time
import zipfile

__all__ = []
__version__ = "1.0.0"  # See https://www.python.org/dev/peps/pep-0396/
__date__ = '2019-03-27'
__updated__ = '2019-11-01'

SENZING_PRODUCT_ID = "5003"  # See https://github.com/Senzing/knowledge-base/blob/master/lists/senzing-product-ids.md
log_format = '%(asctime)s %(message)s'

# Working with bytes.

KILOBYTES = 1024
MEGABYTES = 1024 * KILOBYTES
GIGABYTES = 1024 * MEGABYTES

# The "configuration_locator" describes where configuration variables are in:
# 1) Command line options, 2) Environment variables, 3) Configuration files, 4) Default values

config = {}
configuration_locator = {
    "data_dir": {
        "default": "/opt/senzing/data",
        "env": "SENZING_DATA_DIR",
        "cli": "data-dir"
    },
    "debug": {
        "default": False,
        "env": "SENZING_DEBUG",
        "cli": "debug"
    },
    "g2_dir": {
        "default": "/opt/senzing/g2",
        "env": "SENZING_G2_DIR",
        "cli": "g2-dir"
    },
    "sleep_time_in_seconds": {
        "default": 0,
        "env": "SENZING_SLEEP_TIME_IN_SECONDS",
        "cli": "sleep-time-in-seconds"
    },
    "source_data_dir": {
        "default": "/opt/senzing-source/data/1.0.0",
        "env": "SENZING_SOURCE_DATA_DIR",
        "cli": "source-data-dir"
    },
    "source_g2_dir": {
        "default": "/opt/senzing-source/g2",
        "env": "SENZING_SOURCE_G2_DIR",
        "cli": "source-g2-dir"
    },
    "subcommand": {
        "default": None,
        "env": "SENZING_SUBCOMMAND",
    },
}

# Enumerate keys in 'configuration_locator' that should not be printed to the log.

keys_to_redact = []

# -----------------------------------------------------------------------------
# Define argument parser
# -----------------------------------------------------------------------------


def get_parser():
    ''' Parse commandline arguments. '''

    subcommands = {
        'install': {
            "help": 'Copy source data and g2 directories to a target.',
            "arguments": {
                "--data-dir": {
                    "dest": "data_dir",
                    "metavar": "SENZING_DATA_DIR",
                    "help": "Location of Senzing's data. Default: /opt/senzing/data"
                },
                "--debug": {
                    "dest": "debug",
                    "action": "store_true",
                    "help": "Enable debugging. (SENZING_DEBUG) Default: False"
                },
                "--g2-dir": {
                    "dest": "g2_dir",
                    "metavar": "SENZING_G2_DIR",
                    "help": "Location of Senzing's G2. Default: /opt/senzing/g2"
                },
                "--source-data-dir": {
                    "dest": "source_data_dir",
                    "metavar": "SENZING_SOURCE_DATA_DIR",
                    "help": "Source location of Senzing's data. Default: /opt/senzing-source/data/1.0.0"
                },
                "--source-g2-dir": {
                    "dest": "source_g2_dir",
                    "metavar": "SENZING_SOURCE_G2_DIR",
                    "help": "Source location of Senzing's G2. Default: /opt/senzing-source/g2"
                },
            },
        },
        'package-version': {
            "help": 'Display Senzing Engine version inside docker image.',
            "arguments": {
                "--source-g2-dir": {
                    "dest": "source_g2_dir",
                    "metavar": "SENZING_SOURCE_G2_DIR",
                    "help": "Source location of Senzing's G2. Default: /opt/senzing-source/g2"
                },
            },
        },
        'sleep': {
            "help": 'Do nothing but sleep. For Docker testing.',
            "arguments": {
                "--sleep-time-in-seconds": {
                    "dest": "sleep_time_in_seconds",
                    "metavar": "SENZING_SLEEP_TIME_IN_SECONDS",
                    "help": "Sleep time in seconds. DEFAULT: 0 (infinite)"
                },
            },
        },
        'version': {
            "help": 'Print version of program.',
        },
        'docker-acceptance-test': {
            "help": 'For Docker acceptance testing.',
        },
    }

    parser = argparse.ArgumentParser(prog="senzing-package.py", description="Install Senzing packages. For more information, see https://github.com/Senzing/senzing-package")
    subparsers = parser.add_subparsers(dest='subcommand', help='Subcommands (SENZING_SUBCOMMAND):')

    for subcommand_key, subcommand_values in subcommands.items():
        subcommand_help = subcommand_values.get('help', "")
        subcommand_arguments = subcommand_values.get('arguments', {})
        subparser = subparsers.add_parser(subcommand_key, help=subcommand_help)
        for argument_key, argument_values in subcommand_arguments.items():
            subparser.add_argument(argument_key, **argument_values)

    return parser

# -----------------------------------------------------------------------------
# Message handling
# -----------------------------------------------------------------------------

# 1xx Informational (i.e. logging.info())
# 3xx Warning (i.e. logging.warning())
# 5xx User configuration issues (either logging.warning() or logging.err() for Client errors)
# 7xx Internal error (i.e. logging.error for Server errors)
# 9xx Debugging (i.e. logging.debug())


MESSAGE_INFO = 100
MESSAGE_WARN = 300
MESSAGE_ERROR = 700
MESSAGE_DEBUG = 900

message_dictionary = {
    "100": "senzing-" + SENZING_PRODUCT_ID + "{0:04d}I",
    "130": "Version {0} detected in {1}.",
    "131": "Version {0} detected in Senzing package '{1}'.",
    "132": "Archived {0} to {1}",
    "133": "{0} extracted to {1}",
    "134": "Deleting {0} at version {1}.",
    "135": "{0} deleted.",
    "136": "{0} copied to {1}.",
    "137": "{0} does not exist.",
    "292": "Configuration change detected.  Old: {0} New: {1}",
    "293": "For information on warnings and errors, see https://github.com/Senzing/stream-loader#errors",
    "294": "Version: {0}  Updated: {1}",
    "295": "Sleeping infinitely.",
    "296": "Sleeping {0} seconds.",
    "297": "Enter {0}",
    "298": "Exit {0}",
    "299": "{0}",
    "300": "senzing-" + SENZING_PRODUCT_ID + "{0:04d}W",
    "301": "Cannot determine version. {0} does not exist.",
    "302": "Cannot move {0} to {1}.",
    "303": "Cannot extract {0} to {1}.",
    "304": "Cannot copy {0} to {1}. Error: {2}",
    "305": "Cannot determine version. {0} does not exist.",
    "499": "{0}",
    "500": "senzing-" + SENZING_PRODUCT_ID + "{0:04d}E",
    "695": "Unknown database scheme '{0}' in database url '{1}'",
    "696": "Bad SENZING_SUBCOMMAND: {0}.",
    "697": "No processing done.",
    "698": "Program terminated with error.",
    "699": "{0}",
    "700": "senzing-" + SENZING_PRODUCT_ID + "{0:04d}E",
    "885": "License has expired.",
    "886": "G2Engine.addRecord() bad return code: {0}; JSON: {1}",
    "888": "G2Engine.addRecord() G2ModuleNotInitialized: {0}; JSON: {1}",
    "889": "G2Engine.addRecord() G2ModuleGenericException: {0}; JSON: {1}",
    "890": "G2Engine.addRecord() Exception: {0}; JSON: {1}",
    "891": "Original and new database URLs do not match. Original URL: {0}; Reconstructed URL: {1}",
    "892": "Could not initialize G2Product with '{0}'. Error: {1}",
    "893": "Could not initialize G2Hasher with '{0}'. Error: {1}",
    "894": "Could not initialize G2Diagnostic with '{0}'. Error: {1}",
    "895": "Could not initialize G2Audit with '{0}'. Error: {1}",
    "896": "Could not initialize G2ConfigMgr with '{0}'. Error: {1}",
    "897": "Could not initialize G2Config with '{0}'. Error: {1}",
    "898": "Could not initialize G2Engine with '{0}'. Error: {1}",
    "899": "{0}",
    "900": "senzing-" + SENZING_PRODUCT_ID + "{0:04d}D",
    "999": "{0}",
}


def message(index, *args):
    index_string = str(index)
    template = message_dictionary.get(index_string, "No message for index {0}.".format(index_string))
    return template.format(*args)


def message_generic(generic_index, index, *args):
    index_string = str(index)
    return "{0} {1}".format(message(generic_index, index), message(index, *args))


def message_info(index, *args):
    return message_generic(MESSAGE_INFO, index, *args)


def message_warning(index, *args):
    return message_generic(MESSAGE_WARN, index, *args)


def message_error(index, *args):
    return message_generic(MESSAGE_ERROR, index, *args)


def message_debug(index, *args):
    return message_generic(MESSAGE_DEBUG, index, *args)


def get_exception():
    ''' Get details about an exception. '''
    exception_type, exception_object, traceback = sys.exc_info()
    frame = traceback.tb_frame
    line_number = traceback.tb_lineno
    filename = frame.f_code.co_filename
    linecache.checkcache(filename)
    line = linecache.getline(filename, line_number, frame.f_globals)
    return {
        "filename": filename,
        "line_number": line_number,
        "line": line.strip(),
        "exception": exception_object,
        "type": exception_type,
        "traceback": traceback,
    }

# -----------------------------------------------------------------------------
# Configuration
# -----------------------------------------------------------------------------


def get_configuration(args):
    ''' Order of precedence: CLI, OS environment variables, INI file, default. '''
    result = {}

    # Copy default values into configuration dictionary.

    for key, value in list(configuration_locator.items()):
        result[key] = value.get('default', None)

    # "Prime the pump" with command line args. This will be done again as the last step.

    for key, value in list(args.__dict__.items()):
        new_key = key.format(subcommand.replace('-', '_'))
        if value:
            result[new_key] = value

    # Copy OS environment variables into configuration dictionary.

    for key, value in list(configuration_locator.items()):
        os_env_var = value.get('env', None)
        if os_env_var:
            os_env_value = os.getenv(os_env_var, None)
            if os_env_value:
                result[key] = os_env_value

    # Copy 'args' into configuration dictionary.

    for key, value in list(args.__dict__.items()):
        new_key = key.format(subcommand.replace('-', '_'))
        if value:
            result[new_key] = value

    # Special case: Remove variable of less priority.

    if result.get('project_filespec') and result.get('project_filename'):
        result.pop('project_filename')  # Remove key

    # Special case: subcommand from command-line

    if args.subcommand:
        result['subcommand'] = args.subcommand

    # Special case: Change boolean strings to booleans.

    booleans = [
        'debug'
    ]
    for boolean in booleans:
        boolean_value = result.get(boolean)
        if isinstance(boolean_value, str):
            boolean_value_lower_case = boolean_value.lower()
            if boolean_value_lower_case in ['true', '1', 't', 'y', 'yes']:
                result[boolean] = True
            else:
                result[boolean] = False

    # Special case: Change integer strings to integers.

    integers = [
        'sleep_time_in_seconds'
    ]
    for integer in integers:
        integer_string = result.get(integer)
        result[integer] = int(integer_string)

    return result


def validate_configuration(config):
    ''' Check aggregate configuration from commandline options, environment variables, config files, and defaults. '''

    user_warning_messages = []
    user_error_messages = []

#     if not config.get('g2_database_url'):
#         user_error_messages.append(message_error(401))

    # Perform subcommand specific checking.

    subcommand = config.get('subcommand')

    if subcommand in ['XXX', 'YYY', 'ZZZ']:

        if not config.get('ld_library_path'):
            user_error_messages.append(message_error(414))

    # Log warning messages.

    for user_warning_message in user_warning_messages:
        logging.warning(user_warning_message)

    # Log error messages.

    for user_error_message in user_error_messages:
        logging.error(user_error_message)

    # Log where to go for help.

    if len(user_warning_messages) > 0 or len(user_error_messages) > 0:
        logging.info(message_info(293))

    # If there are error messages, exit.

    if len(user_error_messages) > 0:
        exit_error(697)


def redact_configuration(config):
    ''' Return a shallow copy of config with certain keys removed. '''
    result = config.copy()
    for key in keys_to_redact:
        try:
            result.pop(key)
        except:
            pass
    return result

# -----------------------------------------------------------------------------
# Utility functions
# -----------------------------------------------------------------------------


def create_signal_handler_function(args):
    ''' Tricky code.  Uses currying technique. Create a function for signal handling.
        that knows about "args".
    '''

    def result_function(signal_number, frame):
        logging.info(message_info(298, args))
        sys.exit(0)

    return result_function


def bootstrap_signal_handler(signal, frame):
    sys.exit(0)


def entry_template(config):
    ''' Format of entry message. '''
    debug = config.get("debug", False)
    config['start_time'] = time.time()
    if debug:
        final_config = config
    else:
        final_config = redact_configuration(config)
    config_json = json.dumps(final_config, sort_keys=True)
    return message_info(297, config_json)


def exit_template(config):
    ''' Format of exit message. '''
    debug = config.get("debug", False)
    stop_time = time.time()
    config['stop_time'] = stop_time
    config['elapsed_time'] = stop_time - config.get('start_time', stop_time)
    if debug:
        final_config = config
    else:
        final_config = redact_configuration(config)
    config_json = json.dumps(final_config, sort_keys=True)
    return message_info(298, config_json)


def exit_error(index, *args):
    ''' Log error message and exit program. '''
    logging.error(message_error(index, *args))
    logging.error(message_error(698))
    sys.exit(1)


def exit_silently():
    ''' Exit program. '''
    sys.exit(0)

# -----------------------------------------------------------------------------
# Utility functions
# -----------------------------------------------------------------------------


def get_installed_version(config):
    '''Get version of Senzing seen in senzing_dir.'''

    result = None
    source_g2_dir = config.get('source_g2_dir')
    version_file = "{0}/g2BuildVersion.json".format(source_g2_dir)

    # Read version file.

    try:
        with open(version_file) as version_json_file:
            version_dictionary = json.load(version_json_file)
            result = version_dictionary.get('VERSION')
            logging.info(message_info(130, result, version_file))
    except:
        logging.info(message_warning(305, version_file))

    return result

# -----------------------------------------------------------------------------
# do_* functions
#   Common function signature: do_XXX(args)
# -----------------------------------------------------------------------------


def do_docker_acceptance_test(args):
    ''' For use with Docker acceptance testing. '''

    # Get context from CLI, environment variables, and ini files.

    config = get_configuration(args)

    # Prolog.

    logging.info(entry_template(config))

    # Epilog.

    logging.info(exit_template(config))


def do_install(args):
    '''Install Senzing_API.tgz package. Backup existing version.'''

    # Get context from CLI, environment variables, and ini files.

    config = get_configuration(args)

    source_data_dir = config.get('source_data_dir')
    target_data_dir = config.get('data_dir')
    source_g2_dir = config.get('source_g2_dir')
    target_g2_dir = config.get('g2_dir')

    # Prolog.

    validate_configuration(config)
    logging.info(entry_template(config))

    # Perform action.

    try:
        distutils.dir_util.copy_tree(source_data_dir, target_data_dir, preserve_symlinks=1)
        logging.info(message_info(136, source_data_dir, target_data_dir))
    except Exception as err:
        logging.info(message_warning(304, source_data_dir, target_data_dir, err))

    try:
        distutils.dir_util.copy_tree(source_g2_dir, target_g2_dir, preserve_symlinks=1)
        logging.info(message_info(136, source_g2_dir, target_g2_dir))
    except Exception as err:
        logging.info(message_warning(304, source_g2_dir, target_g2_dir, err))

    # Epilog.

    logging.info(exit_template(config))


def do_package_version(args):
    '''Get version in Senzing_API.tgz package.'''

    # Get context from CLI, environment variables, and ini files.

    config = get_configuration(args)

    # Prolog.

    validate_configuration(config)
    logging.info(entry_template(config))

    # Pull values from configuration.

    get_installed_version(config)

    # Epilog.

    logging.info(exit_template(config))


def do_sleep(args):
    ''' Sleep.  Used for debugging. '''

    # Get context from CLI, environment variables, and ini files.

    config = get_configuration(args)

    # Prolog.

    logging.info(entry_template(config))

    # Pull values from configuration.

    sleep_time_in_seconds = config.get('sleep_time_in_seconds')

    # Sleep.

    if sleep_time_in_seconds > 0:
        logging.info(message_info(296, sleep_time_in_seconds))
        time.sleep(sleep_time_in_seconds)

    else:
        sleep_time_in_seconds = 3600
        while True:
            logging.info(message_info(295))
            time.sleep(sleep_time_in_seconds)

    # Epilog.

    logging.info(exit_template(config))


def do_version(args):
    ''' Log version information. '''

    logging.info(message_info(294, __version__, __updated__))

# -----------------------------------------------------------------------------
# Main
# -----------------------------------------------------------------------------


if __name__ == "__main__":

    # Configure logging. See https://docs.python.org/2/library/logging.html#levels

    log_level_map = {
        "notset": logging.NOTSET,
        "debug": logging.DEBUG,
        "info": logging.INFO,
        "fatal": logging.FATAL,
        "warning": logging.WARNING,
        "error": logging.ERROR,
        "critical": logging.CRITICAL
    }

    log_level_parameter = os.getenv("SENZING_LOG_LEVEL", "info").lower()
    log_level = log_level_map.get(log_level_parameter, logging.INFO)
    logging.basicConfig(format=log_format, level=log_level)

    # Trap signals temporarily until args are parsed.

    signal.signal(signal.SIGTERM, bootstrap_signal_handler)
    signal.signal(signal.SIGINT, bootstrap_signal_handler)

    # Parse the command line arguments.

    subcommand = os.getenv("SENZING_SUBCOMMAND", None)
    parser = get_parser()
    if len(sys.argv) > 1:
        args = parser.parse_args()
        subcommand = args.subcommand
    elif subcommand:
        args = argparse.Namespace(subcommand=subcommand)
    else:
        parser.print_help()
        if len(os.getenv("SENZING_DOCKER_LAUNCHED", "")):
            subcommand = "sleep"
            args = argparse.Namespace(subcommand=subcommand)
            do_sleep(args)
        exit_silently()

    # Catch interrupts. Tricky code: Uses currying.

    signal_handler = create_signal_handler_function(args)
    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)

    # Transform subcommand from CLI parameter to function name string.

    subcommand_function_name = "do_{0}".format(subcommand.replace('-', '_'))

    # Test to see if function exists in the code.

    if subcommand_function_name not in globals():
        logging.warning(message_warning(596, subcommand))
        parser.print_help()
        exit_silently()

    # Tricky code for calling function based on string.

    globals()[subcommand_function_name](args)
